kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,958 @@
+use crate::utilities::cells::{MapCell, OptionalCell, TakeCell};
+use crate::ErrorCode;
+
+const TBF_HEADER_LENGTH: usize = 16;
+pub const BUF_LEN: usize = 512;
+const MAX_PROCS: usize = 10; //need this to store the start addresses of processes to write padding
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum NonvolatileCommand {
+    UserspaceRead,
+    UserspaceWrite,
+}
+
+#[derive(Clone, Copy)]
+pub enum NonvolatileUser {
+    Client,
+    Kernel,
+}
+
+struct NewApp {
+    // App size requested by userland ota app
+    size: usize,
+    // start_addr points the start address where the new app will be loaded
+    start_addr: usize,
+}
+
+/// This interface supports loading processes at runtime.
+pub trait DynamicProcessLoading {
+    /// Call to request loading a new process.
+    ///
+    /// This informs the kernel we want to load a process and the size of the entire process binary.
+    /// The kernel will try to find a suitable location in flash to store said process.
+    ///
+    /// Return value:
+    /// - `Ok((start_address, length))`: If there is a place to load the
+    ///   process, the function will return `Ok()` with the address to start at
+    ///   and the size of the region to store the process.
+    /// - `Err(ErrorCode)`: If there is nowhere to store the process a suitable
+    ///   `ErrorCode` will be returned.
+    fn setup(&self, app_length: usize) -> Result<usize, ErrorCode>;
+
+    /// Instruct the kernel to write data to the flash
+    ///
+    ///This is used to write both userland apps and padding apps
+    fn write_app_data(
+        &self,
+        flag: bool,
+        buffer: &'static mut [u8],
+        offset: usize,
+        app_size: usize,
+        processid: ProcessId,
+    ) -> Result<(), ErrorCode>;
+
+    /// Instruct the kernel to load and execute the process.
+    ///
+    /// This tells the kernel to try to actually execute the new process that
+    /// was just installed in flash from the preceding `write()` call.
+    fn load(&self) -> Result<(), ErrorCode>;
+
+    /// Sets a client for the DynamicProcessLoading Object
+    ///
+    /// When the client operation is done, it calls the
+    /// app_data_write_done(&self, buffer: &'static mut [u8], length: usize) function
+    fn set_client(&self, client: &'static dyn DynamicProcessLoadingClient);
+}
+
+/// The callback for set_client(&self, Client).
+/// The client capsule should implement this trait to handle the callback logic
+pub trait DynamicProcessLoadingClient {
+    fn app_data_write_done(&self, buffer: &'static mut [u8], length: usize);
+}
+
+pub struct DynamicProcessLoader<'a, C: 'static + Chip> {
+    kernel: &'static Kernel,
+    chip: &'static C,
+    fault_policy: &'static dyn ProcessFaultPolicy,
+    procs: MapCell<&'static mut [Option<&'static dyn process::Process>]>,
+    flash: Cell<&'static [u8]>,
+    app_memory: OptionalCell<&'static mut [u8]>,
+    flash_start: Cell<usize>,
+    flash_end: Cell<usize>,
```

Comment:
Why do we need these if we have `self.flash`?

Reply:
Removed these.